### DISCUSSION TITLE
How to run on T-Display-S3

### DISCUSSION THREAD

**Discussion by dapug** [PlummersSoftwareLLC/NightDriverStrip#180] [Q&A]
Has anyone tried this with the LilyGo T-Display-S3?

I made a new section in platformio.ini for this board (very similar to **env:ledstrip_feather**):

```
[env:ttgo-display-s3]
board         = esp32-s3-devkitc-1
monitor_speed = 1000000
upload_speed  = 115200
build_flags   = -DLEDSTRIP=1
                -DUSE_SCREEN=1
                -DUSE_TFTSPI=1
                -DUSER_SETUP_LOADED
                -DTOUCH_CS=0
                -DST7789_2_DRIVER
                -DTFT_WIDTH=170 ; 135                      ; Define the TFT screen setup
                -DTFT_HEIGHT=320 ; 240
                -DTFT_BL=38             
                -DTFT_BACKLIGHT_ON=HIGH
                -DTFT_RGB_ORDER=TFT_RGB              ; Fix color order being swapped
                -DTFT_CS=6                           ; CS, DC, RST                      
                -DTFT_DC=7
                -DTFT_RST=5
                -DTFT_MOSI=11 ; SPI_D                ; MOSI, MISO, CLK
                -DTFT_MISO=13 ; SPI_Q
                -DTFT_SCLK=12 ; SPI_CLK
                -DLOAD_GLCD=1                        ; Font info
                -DLOAD_FONT2=1
                -DLOAD_FONT4=1
                -DLOAD_FONT6=1

                -DLOAD_FONT7=1
                -DLOAD_FONT8=1
                -DLOAD_FONT8N=1
                -DLOAD_GFXFF=1
                -DSMOOTH_FONT=1
                -DSPI_FREQUENCY=27000000                
                -DESP32FEATHERTFT=1                 ; This board has a TFT that is supported by TFT_eSPI
                -DNUM_INFO_PAGES=2
                -std=gnu++17
                -DUSE_PSRAM=1
                -DBOARD_HAS_PSRAM=1
                -mfix-esp32-psram-cache-issue
                -DUNITY_INCLUDE_DOUBLE 
                -DUNITY_DOUBLE_PRECISION=1e-12      ; Make doubles real 8 byte doubles
                -Dregister=                         ; Sinister:  redefined 'register' so FastLED can use that keyword under C++17
                -Os
lib_deps        = ${common.lib_deps}
                  TFT_eSPI
```
Note the correct board as **esp32-s3-devkitc-1**. All pins are configured for this board according to the spec sheet.  Note that it actually uses LEDSTRIP, and in globals.h, and I set LED_PIN0 = 1  (GPIO1 for this board, as Pin 5 is not available for IO)

I also added this in order to enable this board:
```
; Libs needed to link with a TTGO T-Display-S3 Module
[ttgo-display-s3]
lib_deps        = ${common.lib_deps}
                  https://github.com/Xinyuan-LilyGO/T-Display-S3
                  TFT_eSPI
```

Compiles great, writes to device, but does not succeed to run.  I see this in the Serial Monitor terminal:

```
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x2b (SPI_FAST_FLASH_BOOT)
Saved PC:0x40381ff1
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x43c
load:0x403c9700,len:0xbec
load:0x403cc700,len:0x2a3c
entry 0x403c98d8
E (219) psram: PSRAM ID read error: 0x00ffffff
```

I'm sure it's a config or something I missed somewhere.  I've never used this board before but it is very similar to the Adafruit ESP32-S3 TFT, which Dave made official support/config for.  Many thanks to anyone with ideas of what to check (or experience with NightDriver on this particular board)

**Comment by davepl** [comment]
The PSRAM ID error is new to me but sure does sound like a config/setup issue of some kind.  Make sure you DO have PSRAM on that board!

Not sure if this will help, I’m on vacation and away from the code, but here’s an S3 config with PSRAM I just used.  Perhaps useful….

Most of the defines are for the screen on this unit and will need to be removed or updated to reflect the pinout of the display.  But maybe between these two you’ll spot it.



[env:ledstrip_feather]
board         = adafruit_feather_esp32s3_tft
monitor_speed = 1500000
upload_speed  = 1500000
upload_port   = 192.168.8.160
monitor_port  = /dev/cu.usbmodem8411101
build_flags   = -DLEDSTRIP=1
                -DUSE_SCREEN=1
                -DUSE_TFTSPI=1
                -DUSER_SETUP_LOADED
                -DTOUCH_CS=0
                -DST7789_2_DRIVER
                -DTFT_WIDTH=135                      ; Define the TFT screen setup
                -DTFT_HEIGHT=240
                -DTFT_BL=45             
                -DTFT_BACKLIGHT_ON=HIGH
                -DTFT_RGB_ORDER=TFT_RGB              ; Fix color order being swapped
                -DTFT_CS=7                           ; CS, DC, RST                      
                -DTFT_DC=39
                -DTFT_RST=40
                -DTFT_MOSI=35                        ; MOSI, MISO, CLK
                -DTFT_MISO=37
                -DTFT_SCLK=36
                -DLOAD_GLCD=1                        ; Font info
                -DLOAD_FONT2=1
                -DLOAD_FONT4=1
                -DLOAD_FONT6=1
                -DLOAD_FONT7=1
                -DLOAD_FONT8=1
                -DLOAD_FONT8N=1
                -DLOAD_GFXFF=1
                -DSMOOTH_FONT=1
                -DSPI_FREQUENCY=27000000                
                -DESP32FEATHERTFT=1                 ; This board has a TFT that is supported by TFT_eSPI
                -DTOGGLE_BUTTON_1=0                 ; Use GPIO0 button as the info toggle
                -DNUM_INFO_PAGES=2
                -std=gnu++17
                -DUSE_PSRAM=1
                -DBOARD_HAS_PSRAM=1
                -mfix-esp32-psram-cache-issue
                -DUNITY_INCLUDE_DOUBLE 
                -DUNITY_DOUBLE_PRECISION=1e-12      ; Make doubles real 8 byte doubles
                -Dregister=                         ; Sinister:  redefined 'register' so FastLED can use that keyword under C++17
                -Os
lib_deps        = ${common.lib_deps}
                  TFT_eSPI


[env:panlee]
board         = esp32-s3-devkitc-1
board_build.mcu = esp32s3
board_build.f_cpu = 240000000L
upload_speed  = 1000000
monitor_speed = 115200
;upload_port   = /dev/tty.usbmodem84201
monitor_port  = /dev/cu.usbmodem84401
build_flags   = -DLEDSTRIP=1
                -DUSE_SCREEN=1
                -DUSE_TFTSPI=1
                -DUSER_SETUP_LOADED=1
                -DST7796_DRIVER=1
                -DTFT_WIDTH=480
                -DTFT_HEIGHT=320
                -DUSE_HSPI_PORT=1
                -DPIN_SDA=38
                -DPIN_SCL=19
                -DTFT_MISO=12
                -DTFT_MOSI=13
                -DTFT_SCLK=39
                -DTFT_CS=15
                -DTFT_DC=21
                -DTFT_RST=22
                -DTFT_BL=48
                -DLOAD_GLCD
                -DLOAD_FONT2  
                -DLOAD_FONT4
                -DLOAD_FONT6 
                -DLOAD_FONT7
                -DLOAD_FONT8
                -DLOAD_GFXFF
                -DSPI_FREQUENCY=27000000
                -DSPI_READ_FREQUENCY=20000000
                -DSPI_TOUCH_FREQUENCY=2500000
                -DSMOOTH_FONT
                -DNUM_INFO_PAGES=1
                -DPANLEE=1
                -DUSE_PSRAM=1
                -DBOARD_HAS_PSRAM=1
                -mfix-esp32-psram-cache-issue
                -D_IR_ENABLE_DEFAULT_=false  ; don't automatically include every remote control decoder
                -DDECODE_NEC=true            ; enable whichever you need for your remote.  Try not disabling above to figure out which it is.
                -Dregister=                         ; Sinister:  redefine 'register' so FastLED can use that keyword under C++17
                -std=gnu++17
                -O
lib_deps      =  ${common.lib_deps}
                 TFT_eSPI

**Reply by dapug** [reply]
yup, I based my config off the feather, as this board is very similar, and simply updated the pins for this board.

From what I've read, this error is PSRAM ID read error is due to selecting the wrong board (which I didn't, it works fine with other code deployed), or perhaps due to config of the board (namely, PSRAM).  Well, DBOARD_HAS_PSRAM is the official way to enable PSRAM for ESP32 boards.  But I wonder if perhaps the MODE is a factor?  (these boards can use Quad or Octal... but I have no idea how to set that, or whether LilyGo T-Display-S3 PSRAM is different than the Adafruit Feather S3)

Anyway, I disabled PSRAM (comment out ; DBOARD_HAS_PSRAM=1 or set to =0) and the code runs without this error (such as "Demo" rainbow effect).  But I believe that means I can't run as webserver, or do anything I really want that likely requires PSRAM.

Dunno what the deal is.  Will have to dig more later.  Meanwhile... bummer.  

Enjoy your vaca!  Hopefully we can chat one day about MS days (I was in DevDiv on the Windows team back in 99 :) )

**Comment by dapug** [comment]
For anyone following this...

I got NightDriver to load and run LOCAL effects successfully on the LilyGo T-Display-S3, HOWEVER, I cannot get the TFT display or Webserver to run, and I do not know why.  Here is everything I know.

**Why try and get this working on LilyGo T-Display-S3?**

1. Because it is the next evolution from the TTGO T-Display devices (which NightDriver does support)
2. It is very similar to Adafruit Feather S3 TFT... and as of this point in time, they are out of stock and/or hard to find anywhere
3. Because it's a cool new device, readily available from Amazon

### For Local Effects
Using this config, I can get **local effects** running:
```
[env:ttgo-display-s3]
board         = esp32-s3-devkitc-1
monitor_speed = 1000000
upload_speed  = 115200
  
platform_packages =
    framework-arduinoespressif32@https://github.com/espressif/arduino-esp32.git#2.0.5
    
build_flags   = -DDEMO=1
                -std=gnu++17

                -DUSE_PSRAM=1
                -DBOARD_HAS_PSRAM=1
                -mfix-esp32-psram-cache-issue
                -DUNITY_INCLUDE_DOUBLE 
                -DUNITY_DOUBLE_PRECISION=1e-12      ; Make doubles real 8 byte doubles
                -Dregister=                         ; Sinister:  redefined 'register' so FastLED can use that keyword under C++17
                -Os
board_build.partitions = default_8MB.csv ; 
board_build.arduino.memory_type = qio_opi
board_build.psram_type = opi
board_build.flash_size = 8MB
```
IMPORTANT: you also need to set `LED_PIN0 = 1` (or whatever valid GPIO you want)  in `globals.h` (Demo section, or whatever section you intend to run), because Pin 5 is NOT available on this board.

NOTE: this fixes the issue mentioned in my original post (PSRAM ID read error), and gets PSRAM working. The main diff between this and the Adafruit Feather config, for example, is the need to specify `board_build.arduino.memory_type = qio_opi` and `board_build.psram_type = opi`.

### For TFT Display 
According to the developer of t-display-s3, they do NOT actually use SPI when using the TFT_eSPI library (which NightDriver uses for display).  They actually MODIFIED that library to run it in parallel mode. This is supposed to be a **good** thing, as it means much better hardware **perf** (not relying as much on software). BUT... it is problematic when trying to be compatible with NightDriver and pretty much any project out there relying on TFT_eSPI.

Details are outlined here:
https://github.com/Xinyuan-LilyGO/T-Display-S3/issues/9

Some disucssion about how LVGL and i8080 parallel is 30%? faster than TFT_eSPI
https://www.reddit.com/r/esp32/comments/xvyi78/i_made_a_lillygo_tdisplay_s3_setup_guide_with/

I tried running this mod (a few source and config files get tweaked for the t-display-s3) and I had NO success.  I would get this error:
`[E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected`
And I have no idea why, because it doesn't say which pin, and so I cannot track down what to tackle.  What I suspect is happening is that the TFT_eSPI library expects certain pins to be within a range, or even defined at all, and something just isn't lining up, but what?

All my pins are CORRECT for this board, double checked, and also verified with direct samples from the t-display-s3 developer.  Here is what I used for reference (which results in that invalid pin error I cannot seem to shake!)

```
[env:ttgo-display-s3]
board         = esp32-s3-devkitc-1
monitor_speed = 1000000
upload_speed  = 115200
; upload_port   = 192.168.3.160    COM4
; monitor_port  = /dev/cu.usbmodem84401 ;/dev/cu.usbmodem8411101       
platform_packages =
    framework-arduinoespressif32@https://github.com/espressif/arduino-esp32.git#2.0.5
build_flags   = -DDEMO=1
                -DUSE_SCREEN=1
                -DUSE_TFTSPI=1
                -DESP32FEATHERTFT=1                 ; This board has a TFT that is supported by TFT_eSPI
                -DUSER_SETUP_LOADED
                -DTOUCH_CS=0
                -DST7789_DRIVER  ; Note that this is not DST7789_2_DRIVER, because the modified TFT_eSPI lib from t-display-s3 uses DST7789_DRIVER
                -DTFT_WIDTH=172                       ; Define the TFT screen setup
                -DTFT_HEIGHT=320 
                -DTFT_BL=38             
                -DTFT_BACKLIGHT_ON=HIGH
                ; -DTFT_RGB_ORDER=TFT_RGB      ; t-display-s3 actually uses BGR (not RGB) and this is handled in the modified TFT_eSPI lib 
                -DTFT_CS=6                           ; CS, DC, RST                      
                -DTFT_DC=7
                -DTFT_RST=5
                -DTFT_WR=8
                -DTFT_RD=-1 
                ;  NOTE: Not sure why these 3 are not defined in the code by t-display-s3, or whether they are needed, but removing them actually causes MORE "Invalid pin" errors.
                ;        try uncommenting these 3 results in TWO less invalid pin errors... not sure what's with the last and final pin this whole thing is complaining about
                ; -DTFT_MOSI=11 ; SPI_D                ; MOSI, MISO, CLK
                ; -DTFT_MISO=13 ; SPI_Q
                ; -DTFT_SCLK=12 ; SPI_CLK
                -DTOGGLE_BUTTON_1=14                 ; Pin 14 is the "Key" button on the t-display-s3 board
                -DLOAD_GLCD=1                        ; Font info
                -DLOAD_FONT2=1
                -DLOAD_FONT4=1
                -DLOAD_FONT6=1

                -DLOAD_FONT7=1
                -DLOAD_FONT8=1
                ; -DLOAD_FONT8N=1  ; t-display-s3 developer specifically commented this out in their code, not sure wy
                -DLOAD_GFXFF=1
                -DSMOOTH_FONT=1
                ; -DSPI_FREQUENCY=27000000    ; modified TFT_eSPI lib sets this         
                -DNUM_INFO_PAGES=2
                -std=gnu++17
                -DUSE_PSRAM=1
                -DBOARD_HAS_PSRAM=1
                -mfix-esp32-psram-cache-issue
                -DUNITY_INCLUDE_DOUBLE 
                -DUNITY_DOUBLE_PRECISION=1e-12      ; Make doubles real 8 byte doubles
                -Dregister=                         ; Sinister:  redefined 'register' so FastLED can use that keyword under C++17
                -Os
board_build.partitions = default_8MB.csv
board_build.arduino.memory_type = qio_opi
board_build.psram_type = opi
board_build.flash_size = 8MB
lib_deps        = ${common.lib_deps}
                  TFT_eSPI
```

### For Wifi and WebServer

1. Copy include/secrets.example.h as include/secrets.h. Enter your WIFI creds
2. Make sure you enable wifi and webserver in globals.h (in whatever section you are trying to reference in build_flags and run)
3. From the PlatformIO console, upload SPIFFS to your device as mentioned in readme:
    pio run --target buildfs --environment <project name>
    pio run --target uploadfs --environment <project name>

For me, so far, it seems to do literally... nothing.  No errors or logs or activity in the serial monitor, etc.  Just simply, nothing. With the TFT unavailable, I cannot see my IP like I can with my Heltec. Router doesnt show any new devices joined the network.  It's just running my local demo effect (operational but antisocial).  Not sure what the deal is.

No web server or wifi makes me sad. I can live without the TFT display, but sending data to the device is a MUST!
